File: stacks_queues/merge_rods.py
import heapq

def min_cost_rods_merge(arr):
    cost = 0

    min_heap = arr[:]
    heapq.heapify(min_heap)

    while len(min_heap) > 1:
        rod1 = heapq.heappop(min_heap)
        rod2 = heapq.heappop(min_heap)

        cost += rod1 + rod2

        heapq.heappush(min_heap, rod1 + rod2)

    return cost

arr = [1, 2, 3, 2, 1, 4]
print(min_cost_rods_merge(arr))

File: stacks_queues/merge_rods.cpp
#include<bits/stdc++.h>
using namespace std;

int min_cost_rods_merge(vector<int> arr) {
    int cost = 0;

    priority_queue<int, vector<int>, greater<int>> min_heap;
    for (auto x: arr)
        min_heap.push(x);

    while (min_heap.size() > 1) {
        int rod1 = min_heap.top(); min_heap.pop();
        int rod2 = min_heap.top(); min_heap.pop();

        cost += rod1 + rod2;

        min_heap.push(rod1+rod2);
    }

    return cost;
}

int main() {
    vector<int> arr = {1, 2, 3, 2, 1, 4};

    cout << min_cost_rods_merge(arr) << endl;

    return 0;
}